content/blog/tidymodels-sparse-support/index.Rmd

```diff
@@ -0,0 +1,146 @@
+---
+output: hugodown::hugo_document
+slug: tidymodels-sparse-support
+title: Sparse data structures in tidymodels
+date: 2020-11-16
+author: Julia Silge
+description: >
+    Sparse data is common in many domains, and now tidymodels supports using 
+    sparse matrix structures throughout the fitting and tuning stages of modeling.
+photo:
+  url: https://unsplash.com/photos/7JX0-bfiuxQ
+  author: JJ Ying
+# one of: "deep-dive", "learn", "package", "programming", or "other"
+categories: [learn] 
+tags: [tidymodels,tune,parsnip,hardhat]
+
+---
+
+
+```{r setup, include=FALSE}
+library(knitr)
+knitr::opts_chunk$set(cache = TRUE, cache.lazy = FALSE, warning = FALSE, 
+                      message = FALSE, echo = TRUE, dpi = 180,
+                      fig.width = 8, fig.height = 5)
+library(tidyverse)
+theme_set(theme_minimal())
+```
+
+The new release of [tune](linktopost) is chock full of improvements and new features. When combined with the latest releases of [hardhat](http://hardhat.tidymodels.org/) and [parsnip](https://parsnip.tidymodels.org/), one upgrade that tidymodels users can now use in their day-to-day modeling work is **support for sparse data structures**.
+
+## Why sparse data?
+
+In some subject matter domains, it is common to have lots and lots of zeroes after transforming data to a representation appropriate for analysis or modeling. Text data is one such example. The `small_fine_foods` dataset of Amazon reviews of fine foods contains a column `review` that we as humans can read and understand.
+
+```{r R.options=list(quanteda_print_dfm_max_ndoc = 0, quanteda_print_dfm_max_nfeat = 0)}
+library(tidyverse)
+library(tidymodels)
+
+data("small_fine_foods")
+training_data
+```
+
+Computers, on the other hand, need that `review` variable to be heavily preprocessed and transformed in order for it to be ready for most modeling. We typically need to [tokenize](https://smltar.com/tokenization.html) the text, find word frequencies, and perhaps [compute tf-idf](https://www.tidytextmining.com/tfidf.html). We can either keep the resulting data in a long, tidy tibble or we can transform it to a wide matrix, often a good fit when the next step is a modeling or machine learning algorithm.
+
+```{r, R.options=list(quanteda_print_dfm_max_ndoc = 0, quanteda_print_dfm_max_nfeat = 0)}
+library(tidytext)
+
+training_data %>%
+  unnest_tokens(word, review) %>%
+  count(product, word) %>%
+  bind_tf_idf(word, product, n) %>%
+  cast_dfm(product, word, tf_idf)
+```
+
+As is typical for text data, this document-feature matrix is extremely sparse, with many zeroes. Most documents do not contain most words. If we coerce this kind of structure to anything like a `data.frame`, we run into two challenges:
+
+- The amount of **memory** this object requires increases dramatically.
+- We lose out on the **speed** gained from any specialized model algorithms built for sparse data.
+
```

Comment:
I'd briefly mention how sparse matrices work under the hood and maybe do an `obj_size()` calculation with it and its data frame version.